### ISSUE TITLE
[MPS] F.conv3d runs at about 7% of conv2d throughput on the same hardware

### ISSUE THREAD

**Issue by 4suino** [pytorch/pytorch#192213]
### Description

On Apple Silicon, `F.conv3d` reaches roughly 7% of the throughput the same
machine achieves on `F.conv2d`, measured in FLOP/s on shapes with identical
arithmetic intensity. `conv2d` and `matmul` both run at the hardware's
practical peak, so this is specific to the Conv3d path.

The decisive case is the last row below: a Conv3d with a **temporal kernel of
1** is arithmetically a Conv2d applied per frame, and it is still an order of
magnitude slower. The loss is therefore in the Conv3d dispatch rather than in
3D arithmetic.

### Measurements

| Operation | TFLOP/s | Share of matmul |
|---|---:|---:|
| matmul 8192^3 fp16 | 61.41 | 100% |
| conv2d 512->512 k3 | 62.26 | 101% |
| conv3d 512->512 k3x3x3 | 4.12 | **7%** |
| conv3d, temporal kernel 1 | 4.61 | **8%** |

### Reproducer

```python
import statistics, time, torch
import torch.nn.functional as F

d = torch.device("mps")

def bench(fn, n=3):
    for _ in range(2):
        fn()
    torch.mps.synchronize()
    s = []
    for _ in range(n):
        torch.mps.synchronize()
        t0 = time.perf_counter()
        fn()
        torch.mps.synchronize()
        s.append(time.perf_counter() - t0)
    return statistics.median(s)

N = 8192
a = torch.randn(N, N, device=d, dtype=torch.float16)
t = bench(lambda: a @ a)
peak = 2 * N ** 3 / t / 1e12
print(f"matmul                        {peak:6.2f} TFLOP/s  100%")

x2 = torch.randn(1, 512, 512, 512, device=d, dtype=torch.float16)
w2 = torch.randn(512, 512, 3, 3, device=d, dtype=torch.float16)
t = bench(lambda: F.conv2d(x2, w2))
tf = 2 * 512 * 9 * (512 * 510 * 510) / t / 1e12
print(f"conv2d 512->512 k3            {tf:6.2f} TFLOP/s  {tf/peak*100:4.0f}%")

x3 = torch.randn(1, 512, 7, 130, 514, device=d, dtype=torch.float16)
w3 = torch.randn(512, 512, 3, 3, 3, device=d, dtype=torch.float16)
t = bench(lambda: F.conv3d(x3, w3))
tf = 2 * 512 * 27 * (512 * 5 * 128 * 512) / t / 1e12
print(f"conv3d 512->512 k3x3x3        {tf:6.2f} TFLOP/s  {tf/peak*100:4.0f}%")

w3t = torch.randn(512, 512, 1, 3, 3, device=d, dtype=torch.float16)
t = bench(lambda: F.conv3d(x3, w3t))
tf = 2 * 512 * 9 * (512 * 7 * 128 * 512) / t / 1e12
print(f"conv3d, temporal kernel 1     {tf:6.2f} TFLOP/s  {tf/peak*100:4.0f}%")
```

### Practical impact

Expressing the same convolution manually as temporal Conv2d slices, summed, is
about 12x faster at the operator and roughly 5x faster end to end on a video
VAE where Conv3d accounts for 93% of decode time. Same arithmetic, one twelfth
of the time. Any 3D convolutional model on this backend is likely paying this
silently.

### Notes

Conv3d support was added in #114183, which documents that channels_last is not
supported for 3D. Measured here, `channels_last_3d` is 13% slower than the
default layout, consistent with there being no fast path for it.

Existing MPS Conv3d issues cover missing ops (#77818, #117949) and correctness
(#142836), but I could not find one on throughput.

### Environment

PyTorch 2.12.0, Python 3.12.11, macOS on Apple M5 Max (40-core GPU).

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### Why this matters in practice

I maintain a local video restoration app on Apple Silicon built around SeedVR2,
whose VAE is 3D-convolutional. Profiling decode attributes about 93% of the
time to `conv3d`, so this single operator is essentially the entire cost of the
product. Expressing the same convolution manually as temporal Conv2d slices
took a 45-frame 1440p job from 11 minutes to 2 on the same machine, with
outputs agreeing to one ULP.

That is the order of magnitude at stake for anyone running a video model on
this backend, and I suspect most people are paying it without knowing, because
the failure mode is "slow" rather than "wrong".

**Comment by Isalia20** [comment]
This should be addressed by https://github.com/pytorch/pytorch/pull/188802

Can you try the nightly version of torch and see if you get better performance?

**Comment by 4suino** [comment]
Thanks — tested `2.14.0.dev20260804` on the same machine (M5 Max, macOS 26.5.2).
Same script as the issue body, nightly against the 2.12.0 numbers:

| Operation | 2.12.0 | 2.14.0.dev20260804 |
|---|---:|---:|
| matmul 8192^3 fp16 | 61.41 TFLOP/s | 61.14 |
| conv2d 512->512 k3 | 62.26 (101%) | 61.03 (100%) |
| conv3d 512->512 k3x3x3 | 4.12 (**7%**) | 5.40 (**9%**) |
| conv3d, temporal kernel 1 | 4.61 (8%) | 6.23 (10%) |

So about +31% on `conv3d`, but the gap against `conv2d` on the same hardware is
still roughly an order of magnitude rather than closed.

Expressing the same convolution manually as temporal Conv2d slices is still
faster on the nightly, on real decoder shapes, float16:

| Shape | `conv3d` | manual Conv2d slices | |
|---|---:|---:|---:|
| (1, 512, 7, 130, 514) | 850.5 ms | 86.0 ms | 9.89x |
| (1, 128, 7, 258, 1026) | 168.9 ms | 26.3 ms | 6.42x |
| (1, 256, 7, 130, 1026) | 378.6 ms | 47.1 ms | 8.03x |
| (1, 512, 5, 258, 258) | 509.9 ms | 58.7 ms | 8.69x |

Outputs agree to one ULP: max abs difference 0.500 against an output magnitude
of 726.0, which is exactly the float16 granularity there.

Separately, and worth recording: **the INT_MAX limit is fixed**. On 2.12 a
`conv3d` whose activation exceeds 2^31 elements aborts inside MPSGraph with a
negative flattened extent:

```
'mps.reshape' op the result shape is not compatible with the input shape
%5 = "mps.reshape"(%2, %4) : (tensor<-48234496xf16>, tensor<5xsi32>) -> tensor<1x512x9x720x1280xf16>
failed assertion `original module failed verification'
```

On the nightly the same shapes succeed:

| Shape | Elements | vs 2^31 | 2.12.0 | nightly |
|---|---:|---:|---|---|
| (1, 256, 9, 720, 1280) | 2,123,366,400 | 0.99x | — | OK |
| (1, 512, 9, 720, 1280) | 4,246,732,800 | 1.98x | assertion failure | OK |

`conv3d_simd_long` appears to do exactly what it says. That removes a real
constraint downstream: it is why the VAE in our case has to be spatially tiled
at all.

**Comment by Isalia20** [comment]
are you a real person or a bot?

**Comment by 4suino** [comment]
I'm real :) Just translating in English why?

**Comment by 4suino** [comment]
Just to be clear. Happy to be explicit about the workflow, since it's a fair question given the turnaround: I use an AI assistant to run the benchmarks and draft the write-ups, then I review and post. I don't think that's a problem here, everything is a reproducer you can run yourself, and that's the part that should carry the weight rather than who typed it. If the project would rather I flag that on each comment, say so and I will.

Context for why I'm pushing on this. I maintain a local video restoration app on Apple Silicon built around SeedVR2, whose VAE is 3D-convolutional. "conv3d" is about 93% of decode time there, so this single operator is essentially the entire cost of the product. Routing around it took a 45-frame 1440p job from 11 minutes to 2 on the same machine. That is the order of magnitude at stake for anyone running video models on this backend, and I suspect most people are paying it without knowing, since the failure mode is "slow" rather than "wrong".

Where I'm struggling is telling what is already settled. #188802 is a good example: it is closed rather than merged in the UI, and I only worked out that it had actually landed by checking that "Convolution.metal" and "conv3d_simd_long" exist on main. From outside it is hard to tell which MPS performance work is in flight, which has landed but not yet reached a release, and which known gaps are considered acceptable versus simply unclaimed.

So, concretely: is there a label, a tracking issue or a project board that shows current MPS performance work? I'd rather test against what you already have in progress and report back than open issues for things someone is halfway through fixing.

Happy to keep running measurements on this hardware if that's useful - M5 Max, macOS 26.5.2, and I can turn a benchmark around quickly.

**Comment by Isalia20** [comment]
> Where I'm struggling is telling what is already settled. https://github.com/pytorch/pytorch/pull/188802 is a good example: it is >>> closed rather than merged in the UI, and I only worked out that it had actually landed by checking that "Convolution.metal" and >>> "conv3d_simd_long" exist on main. From outside it is hard to tell which MPS performance work is in flight, which has landed but >>> not yet reached a release, and which known gaps are considered acceptable versus simply unclaimed.


All pytorch PRs are closed and merged with @pytorchbot , label is assigned of `Merged`. For what is in works, you can take a look at the open/draft PRs which have label of `ciflow/mps`. As for this issue, I'll take a deeper look later today

**Comment by Isalia20** [comment]
As for AI generated issues, please read https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md

**Comment by 4suino** [comment]
> As for AI generated issues, please read https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md

Understood, thanks.

**Comment by Isalia20** [comment]
>Context for why I'm pushing on this. I maintain a local video restoration app on Apple Silicon built around SeedVR2, whose VAE is 3D-convolutional. "conv3d" is about 93% of decode time there, so this single operator is essentially the entire cost of the product. Routing around it took a 45-frame 1440p job from 11 minutes to 2 on the same machine. That is the order of magnitude at stake for anyone running video models on this backend, and I suspect most people are paying it without knowing, since the failure mode is "slow" rather than "wrong".

For example this above is a great context for the issue, as it shows improvement to this will have real value for you (and for others having same workflow) and would be amazing if you included this in the main issue

**Comment by Isalia20** [comment]
This should fix the issue, thanks for the report:
https://github.com/pytorch/pytorch/pull/192229

**Comment by 4suino** [comment]
> This should fix the issue, thanks for the report: [#192229](https://github.com/pytorch/pytorch/pull/192229)

I'm checking it now. Thanks for the report!

**Comment by 4suino** [comment]
Checked it, and sorry, I have to correct my own numbers first. They were misleading and I am sorry about that.

Your PR made me realise my reproducer calls `F.conv3d(x, w)` with no bias. So everything I measured on the nightly was the SIMD fallback, not the MPP path. I re-ran it on the same machine with a bias term, which is what the model I actually care about uses (all 35 conv3d layers in that VAE have bias):

> | | no bias | with bias |
> |---|---:|---:|
> | 2.12.0 | 4.11 TFLOP/s (7%) | 3.63 TFLOP/s (6%) |
> | 2.14.0.dev20260804 | 5.40 TFLOP/s (9%) | **59.13 TFLOP/s (97%)** |
>
> matmul peak on this machine is about 61 TFLOP/s. Shape (1, 512, 7, 130, 514), fp16, kernel 3x3x3.

So on the nightly the bias case is already at parity with matmul and conv2d. My "still an order of magnitude behind" was simply wrong, it only described the path your PR is now fixing.

On 2.12 the bias makes no difference, both are around 7%, which I assume is because the Metal kernels are not there at all in that release. That part of the original report still holds for 2.12.

One more thing that might be useful to you. With bias on the nightly, native `conv3d` is now faster than the manual Conv2d decomposition on most of my real shapes:

> | Shape | native conv3d | manual conv2d slices |
> |---|---:|---:|
> | (1, 512, 7, 130, 514) | 80.3 ms | 89.8 ms |
> | (1, 128, 7, 258, 1026) | 21.2 ms | 33.2 ms |
> | (1, 512, 5, 258, 258) | 64.4 ms | 53.1 ms |
> | (1, 256, 7, 130, 1026) | 40.0 ms | 52.3 ms |

Which means once this lands I can drop my workaround entirely. That is a good outcome for me, thanks for looking at it so quickly.

Measurements above were run by me with an AI assistant helping me in the benchmarks, on M5 Max, macOS 26.5.2.

**Comment by Isalia20** [comment]
Thank you for testing it out and for the report
